### ISSUE TITLE
Ability to rename and create files

### ISSUE THREAD

**Issue by japgolly** [scalacenter/scalafix#1469]
I'd love to write two scalafix rules:

1. split scala files with multiple top-level types in to one file per type. (eg. `lib.scala` => `Hi.scala` and `Ok.scala`)
2. move files so that their path matches the package (i.e. Java style)

I don't see any functionality that I could use to rename or create files. I was looking through all of the `Patch` methods and there was nothing there. Is there some other means to accomplish this?

**Comment by bjaglin** [comment]
Indeed, there is nothing built-in in Scalafix for that. I am not aware of any previous (public) discussion either.

(2) could easily be implemented by an addition to the `Patch` API.

(1) would require much more thinking. Spontaneously, I feel like a Fork/Copy patch could be the easiest to implement and reason about (in terms of patch composeability and avoiding generating broken source because of missing imports). This patch could itself embed other patches to be applied on the forked file to allow further processing.

**Comment by japgolly** [comment]
Hi again,

So for now, I've made my own app called [scala-restructure](https://github.com/japgolly/scala-restructure) that does what I need it to do, but directly using scalameta, rather that trying to use scalafix and the Patch API. It works well and is a great alternative (especially considering that scalafix is OSS) but if anyone has the time and inclination, it would be awesome to find a way to integrate that work into scalafix so that these could be normal scalafix rules.

Anyway, I hope that's helpful. Thanks!

**Comment by sake92** [comment]
I also had similar ideas, to generate scala files and later update them as needed.  
Made a library for exactly that: https://github.com/sake92/regenesca
Example: https://github.com/sake92/regenesca/blob/main/example/src/example/Example.scala

Used them for 2 (re)generators:
- squery SQL models/daos https://sake92.github.io/squery/tutorials/code-gen.html
- openapi generator https://github.com/sake92/openapi4s

There is certainly an amounr of ad-hoc rules for replacement of definitions, but for my usecases it works surprisingly well.  
And mostly simple to understand, not a lot of code really.
